Design review, not for merge ---  Read source context file

Deren-Liao opened the pull request:
This is alpha version code to help approve the design idea.  Not final code for merge. 
Please comment on "overall design"  like
  + Should we put the code in this lib
  + Is it Okay to introduce dependencies from app domain .
Please ignore
  + comment
  + typo

The source context file is placed at the [root] of the deployment.  It will be used for Cloud Debugger,  Stackdriver Logging, Stackdriver Error Reporting etc. 
I am adding the "reader" into Gax so that all the libs can reference and use it.

Deren-Liao committed:
Add SourceContext

Deren-Liao committed:
Merge branch 'master'

Conflicts:
	src/Google.Api.Gax/project.json

ivannaranjo commented on src/Google.Api.Gax/SourceContext.cs:
```diff
@@ -0,0 +1,90 @@
+        /// </summary>
+        /// <exception cref="AppDomainUnloadedException">Possibly threw by getting base directory.</exception>
+        private static string SourceContextFilePath => 
+            Path.Combine(AppDomain.CurrentDomain.BaseDirectory, SourceContextFileName);
```
Isn't `System.Reflection.Assembly.GetEntryAssembly().Location` a better choice to get the path of the process? It also doesn't need the unofficial `System.AppDomain` package below.

Deren-Liao replied:
Looks it's a better choice. I'll try and see how it works.

Deren-Liao replied:
System.Reflection.Assembly.GetEntryAssembly() returns null on my MVC project.

Commit:
Script to regenerate protos

Commit:
Regenerated protos, using 3.2.0

(Mostly whitespace, as protoc handles comments better now.)

Commit:
Update protobuf dependency to 3.2.0

(Removed explicit dependency in Google.Api.Gax.Grpc - we depend on CommonProtos,
so we don't need it.)

Commit:
Update dependencies

Google.Api.Gax - and therefore everything downstream of it - now depends on NETStandard.Library.
See http://stackoverflow.com/questions/42946951 for background.

Additionally, Google.Api.Gax.Rest depends on System.Net.Http, so we don't need
to restate the dependency in Google.Api.Gax.Rest.Tests

chrisdunelm committed:
Upgrade Newtonsoft.Json from v9.0.1 -> v10.0.1 (#158)

This is required because v9 mishandles u64 numbers greater than i64.max when decoding
Fixes #155

Deren-Liao committed:
Add SourceRevision

Deren-Liao committed:
s

Deren-Liao committed:
merge fix project.json
